types/src/ledger_info.rs

```diff
@@ -373,6 +374,192 @@ impl LedgerInfoWithPartialSignatures {
+    }
+
+    pub fn aggregate_and_verify(
+        &mut self,
+        verifier: &ValidatorVerifier,
+    ) -> Result<LedgerInfoWithSignatures, VerifyError> {
+        self.check_voting_power(verifier)?;
+
+        // Compute signatures from verified signatures and unverified signatures that are not already
+        // included in the verified aggregate signature.
+        let mut partial_signatures = self.verified_signatures.clone();
+        for (author, signatures) in self.unverified_signatures.iter() {
+            partial_signatures.add_signature(*author, signatures[0].clone());
+        }
+        let authors_in_aggregated_sig = self.verified_aggregate_signature.get_signers_addresses(
+            verifier
+                .get_ordered_account_addresses_iter()
+                .collect_vec()
+                .as_slice(),
+        );
+        for author in authors_in_aggregated_sig {
+            partial_signatures.remove_signature(author);
+        }
+
+        if partial_signatures.is_empty() {
+            return Ok(LedgerInfoWithSignatures::new(
+                self.ledger_info.clone(),
+                self.verified_aggregate_signature.clone(),
+            ));
+        }
+
+        let aggregated_sig = verifier.aggregate_new_signatures_with_existing(
+            &partial_signatures,
+            Some(self.verified_aggregate_signature.clone()),
+        )?;
+        let verified_aggregate_signature =
+            match verifier.verify_multi_signatures(self.ledger_info(), &aggregated_sig) {
+                Ok(_) => aggregated_sig,
+                Err(_) => {
+                    let verified = self
+                        .unverified_signatures
+                        .clone()
+                        .into_par_iter()
+                        .flat_map(|(account_address, signatures)| {
+                            // Get the first signature that verifies correctly.
+                            for signature in signatures {
+                                if verifier
+                                    .verify(account_address, self.ledger_info(), &signature)
+                                    .is_ok()
+                                {
+                                    return Some((account_address, signature));
+                                }
+                            }
+                            None
+                        })
+                        .collect::<Vec<_>>();
+                    for (account_address, signature) in verified {
+                        self.verified_signatures
+                            .add_signature(account_address, signature);
+                    }
+                    let aggregated_sig = verifier.aggregate_new_signatures_with_existing(
+                        &self.verified_signatures,
+                        Some(self.verified_aggregate_signature.clone()),
+                    )?;
+                    verifier.verify_multi_signatures(self.ledger_info(), &aggregated_sig)?;
```

Comment:
I don't think this is what you expect, a partial aggregated signature can't pass the verification because we check for voting power as well.

Reply:
I see. Will fix it.